### ISSUE TITLE
Sitemap LastMod property needs to be set properly

### ISSUE THREAD

**Issue by thabaum** [oqtane/oqtane.framework#2764]
When adding, updating, moving it, or deleting a page module the Page.ModifiedOn property needs to update to reflect the date of changes in the Sitemap.

I have a PR for this at the finish line.

**Comment by sbwalker** [comment]
The ModifiedOn and ModifiedBy properties have very specific usage and meaning in regards to entities within the framework. They retain audit information related to the date/time and user who last updated a specific entity These properties should not be used for other purposes or else they lose their context and meaning. 

Tracking the last date/time when a piece of content was updated on a "page" within a site is a much more complex scenario as it needs to take into consideration all of the content dependencies for that page.  This includes any changes to the page properties, module properties, or module content... as if any of these change, it means that the "page" has been modified and should be re-indexed by the search crawler (which is the purpose of the LastMod property in the SiteMap).

This will require a new column to be added to the database specifically for this purpose. It will also require an API method which module developers can use to set the property.  It will also likely need some sort of data migration logic to initialize the property for existing sites. In summary it will require a fair amount of effort to implement so it requires an evaluation if the end result is worth the effort at this time. 

A much simpler solution would be to set the LastMod in the SiteMap to the current date - which would indicate to the search crawler that it should always re-index the content. I have my suspicions that search crawlers do not rely on this metadata anyways, because it is often unreliable - so they use their own custom logic to optimize for efficiency in indexing.

**Comment by thabaum** [comment]
`                    writer.WriteElementString("lastmod", url.ModifiedOn.ToString("yyyy-MM-dd"));`

So are you suggesting changing the above to:

`                    writer.WriteElementString("lastmod", DateTime.UtcNow.ToString("yyyy-MM-dd"));`

in sitemap.cshtml.cs?

**Comment by thabaum** [comment]
I did another where created an API solution as well adding another method to call named UpdatePageModifiedOn, however I was trying not to write any new code, use existing.  I did ask if there was a method that would do this currently but removed that comment.

```
            _db.Entry(page).State = EntityState.Modified;
            _db.SaveChanges();
```
            
            This was the code I was interested in to perform such a task.  I was trying to stay strictly in the PageController.cs and PageModuleController.cs. so I used the `UpdatePage()` function for some feedback here.  Which I got thanks!
            
            Also I meant to remove the _logger and was only using `UpdatePage()` method as it was not needed, but at the time I was playing with those features and didnt remove them to hear more about them ;) but was going to adjust it.
            
            I will have to look but I may have that method I created for the API.  I was just trying to get feedback to see what was an acceptable solution since, well, this text chat stuff here is dull.  
            
            The methods in the controller would of looked like this:
            
```
                    // POST api/<controller>
        [HttpPost]
        [Authorize(Roles = RoleNames.Registered)]
        public PageModule Post([FromBody] PageModule pageModule)
        {
            var page = _pages.GetPage(pageModule.PageId);
            if (ModelState.IsValid && page != null && page.SiteId == _alias.SiteId && _userPermissions.IsAuthorized(User, page.SiteId, EntityNames.Page, pageModule.PageId, PermissionNames.Edit))
            {
                pageModule = _pageModules.AddPageModule(pageModule);
                _pages.UpdatePageModifiedOnAndModifiedBy(page);
                _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.PageModule, pageModule.PageModuleId, SyncEventActions.Create);
                _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.Site, _alias.SiteId, SyncEventActions.Refresh);
                _logger.Log(LogLevel.Information, this, LogFunction.Create, "Page Module Added {PageModule}", pageModule);
            }
            else
            {
                _logger.Log(LogLevel.Error, this, LogFunction.Security, "Unauthorized PageModule Post Attempt {PageModule}", pageModule);
                HttpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
                pageModule = null;
            }
            return pageModule;
        }

        // PUT api/<controller>/5
        [HttpPut("{id}")]
        [Authorize(Roles = RoleNames.Registered)]
        public PageModule Put(int id, [FromBody] PageModule pageModule)
        {
            var page = _pages.GetPage(pageModule.PageId);
            if (ModelState.IsValid && page != null && page.SiteId == _alias.SiteId && _pageModules.GetPageModule(pageModule.PageModuleId, false) != null && _userPermissions.IsAuthorized(User, page.SiteId, EntityNames.Page, pageModule.PageId, PermissionNames.Edit))
            {
                pageModule = _pageModules.UpdatePageModule(pageModule);
                _pages.UpdatePageModifiedOnAndModifiedBy(page);
                _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.PageModule, pageModule.PageModuleId, SyncEventActions.Update);
                _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.Site, _alias.SiteId, SyncEventActions.Refresh);
            }
            else
            {
                _logger.Log(LogLevel.Error, this, LogFunction.Security, "Unauthorized PageModule Put Attempt {PageModule}", pageModule);
                HttpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
                pageModule = null;
            }
            return pageModule;
        }

        // PUT api/<controller>/?pageid=x&pane=y
        [HttpPut]
        [Authorize(Roles = RoleNames.Registered)]
        public void Put(int pageid, string pane)
        {
            var page = _pages.GetPage(pageid);
            if (page != null && page.SiteId == _alias.SiteId && _userPermissions.IsAuthorized(User, page.SiteId, EntityNames.Page, pageid, PermissionNames.Edit))
            {
                int order = 1;
                List<Models.PageModule> pagemodules = _pageModules.GetPageModules(page.SiteId)
                    .Where(item => item.PageId == pageid && item.Pane == pane).OrderBy(item => item.Order).ToList();
                foreach (Models.PageModule pagemodule in pagemodules)
                {
                    if (pagemodule.Order != order)
                    {
                        pagemodule.Order = order;
                        _pageModules.UpdatePageModule(pagemodule);
                        _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.PageModule, pagemodule.PageModuleId, SyncEventActions.Update);
                    }
                    order += 2;
                }
                _pages.UpdatePageModifiedOnAndModifiedBy(page);
                _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.Site, _alias.SiteId, SyncEventActions.Refresh);
                _logger.Log(LogLevel.Information, this, LogFunction.Update, "Page Module Order Updated {PageId} {Pane}", pageid, pane);
            }
            else
            {
                _logger.Log(LogLevel.Error, this, LogFunction.Security, "Unauthorized PageModule Put Attempt {PageId} {Pane}", pageid, pane);
                HttpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
           }
        }

        // DELETE api/<controller>/5
        [HttpDelete("{id}")]
        [Authorize(Roles = RoleNames.Registered)]
        public void Delete(int id)
        {
            PageModule pagemodule = _pageModules.GetPageModule(id);
            if (pagemodule != null && pagemodule.Module.SiteId == _alias.SiteId && _userPermissions.IsAuthorized(User, pagemodule.Module.SiteId, EntityNames.Page, pagemodule.PageId, PermissionNames.Edit))
            {
                _pageModules.DeletePageModule(id);
                _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.PageModule, pagemodule.PageModuleId, SyncEventActions.Delete);
                _syncManager.AddSyncEvent(_alias.TenantId, EntityNames.Site, _alias.SiteId, SyncEventActions.Refresh);
                _logger.Log(LogLevel.Information, this, LogFunction.Delete, "Page Module Deleted {PageModuleId}", id);

                Page page = _pages.GetPage(pagemodule.PageId);
                if (page != null)
                {
                    _pages.UpdatePageModifiedOnAndModifiedBy(page);
                }
            }
            else
            {
                _logger.Log(LogLevel.Error, this, LogFunction.Security, "Unauthorized PageModule Delete Attempt {PageModuleId}", id);
                HttpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
            }
        }
```

`_pages.UpdatePageModifiedOnAndModifiedBy(page);` logic would handle whatever concerns

Also a `_pages.UpdatePageModifiedOn(page);` method in the API Controller and Repository

I made a few versions.  I used to look at Windows OS files thinking I could make an OS one day bootstrapping... learning as I go but currently a one person team here so doing what I can.  

I had also thought of adding another property and having options for Sitemap in the Admin Panel.  I don't want to do any overkill and I was trying to keep it simple as can be.  UpdatePage pulled it off since that is what was being done, but I was concerned with the rest of the logic as well such as setting permissions which wasn't needed.

I think we are getting down to the bone here.  Simple fix is as you said since Search Engines really dont anything but AI Algorithms and bots.  I love how google would bot pages that are not on sitemap and didnt exist in DNN causing my error logs to flood.  Annoying.  It is like they are pretty reckless more than doing what they are suppose to do according to Sitemap.org protocols.

**Comment by thabaum** [comment]
Would it be OK to introduce a new controller and interface to help separate concerns and maintainability?

PageModuleSitemapController, IPageModuleSitemapRepository, PageModuleSitemapRepository and add LastMod parameter to Page model class?

**Comment by sbwalker** [comment]
My preference at this time is to implement the simpler solution outlined above - simply set the LastMod to the current date/time in the SiteMap generated. This accomplishes the primary goal of ensuring content is indexed and has the minimum impact on the framework.

**Comment by sbwalker** [comment]
resolved in #2747
